viewing_party/party.py

```diff
@@ -1,23 +1,153 @@
+    user_data['watchlist'].append(movie)
+    return user_data
+
+def watch_movie(user_data, title):
+    for movie_dict in user_data['watchlist']:
+        if title == movie_dict['title']:
+            user_data['watched'].append(movie_dict)
+            user_data['watchlist'].remove(movie_dict)
+
+    return user_data
+
+    
+
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    total_rating = 0.0
+    avg_rating = 0.0
+    for movie_dict in user_data['watched']:
+        total_rating += movie_dict['rating']
+        if not len(user_data['watched']):
+            avg_rating = 0.0
+        else:
+            avg_rating = total_rating / len(user_data['watched'])
+    return avg_rating
+            
+
 
 
+def get_most_watched_genre(user_data):
+    genre_count = {}
+    if user_data['watched'] == []:
+        return None
+    for movie_dict in user_data['watched']:
+        current_genre = movie_dict['genre']
+        if current_genre not in genre_count:
+            genre_count[current_genre] = 1
+        else:
+            genre_count[current_genre] += 1
+    genre_most_watched = max(genre_count, key = genre_count.get)
+    return genre_most_watched
+    
+    
+        
+    
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_friends_movies(user_data):
+    friends_movies = []
+    for friend in user_data['friends']:
+        for movie in friend['watched']:
+            friends_movies.append(movie)
+    return friends_movies
+
+
+def get_unique_watched(user_data):
+    unique_movies = []
+    friends_movies = get_friends_movies(user_data)
+    for movie_dict in user_data['watched']:
+        if movie_dict not in friends_movies:
+            unique_movies.append(movie_dict)
+    return unique_movies
+
+def get_friends_unique_watched(user_data):
+    user_watched_list = []
+    friends_unique_movies = []
+    index = 0
+    for movies_dict in user_data['watched']:
+        user_watched_list.append(movies_dict['title'])
+
+    for friends_dict in user_data['friends']:
+        movie_list = friends_dict['watched']
+        for movies_dict in movie_list:
+            if movies_dict in friends_unique_movies:
+                continue
+            if movies_dict['title'] not in user_watched_list:
+                friends_unique_movies.append(movies_dict)
+    return friends_unique_movies
+
+    
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+    friends_movies = get_friends_unique_watched(user_data)
+    subscriptions = user_data['subscriptions']
+    reccomendations = []
+
+    for movie_dict in friends_movies:
+        if movie_dict['host'] in subscriptions:

```

Comment:
What's the time complexity of a for loop that has to execute the `in` operator on a list during every iteration?

We could improve the subscription lookup if we trade a little space and make a set of the subscriptions (which are strings, so can go in a set). Using in on a list is a linear operation, but on a set, it's constant.

```python
subscription_lookup = set(user_data["subscriptions"]) 
for movie in friends_watched: 
    if movie["host"] in subscription_lookup: 
        recommended.append(movie)
```